### ISSUE TITLE
Shows false negative error (visual only) for an attribute of an object being assigned to a variable via function return

### ISSUE THREAD

**Issue by ARKaraoglu** [microsoft/pyright#8404]
**Describe the bug**
I am trying to assign the return value of a function into a variable for testing purposes. Pyright shows that it is an error but whenever I print the object and run the tests, the tests all pass without any errors. It seems that its just a visual bug only.

When I opened the same project in PyCharm, PyCharm did not seem to give me any errors.

**Code or Screenshots**
![leafnode](https://github.com/user-attachments/assets/4b34c900-2860-4f9b-9898-82f3496122b6)
![ErrorImage](https://github.com/user-attachments/assets/4c7cabcd-0692-4e76-ba88-5c58b3b58a18)

```python
import unittest

from textnode import TextNode
from leafnode import LeafNode
from leafnode import text_node_to_html_node

class TestTextNode(unittest.TestCase):
    def test_eq(self):
        # URL Empty
        node = TextNode("This is a text node", "bold", "")    
        node2 = TextNode("This is a text node", "bold", "")     
        self.assertEqual(node, node2)
        
        #text_type different
        node3 = TextNode("this is a text node", "bold", "/")
        node4 = TextNode("this is a text node", "", "/")
        self.assertNotEqual(node3, node4)

        #text different
        node5 = TextNode("this is a text node", "bold", "/")
        node6 = TextNode("this is also a text node", "bold", "/")
        self.assertNotEqual(node5, node6)
    
    #node_types = ["text", "bold", "italic", "code", "link", "image"]
    def test_textnode_to_leafnode_text(self):
        textnode1 = TextNode(text = "textnode1", text_type="text", url="")
        leafnode1 = text_node_to_html_node(textnode1)
        
        self.assertIsInstance(leafnode1, LeafNode)
        self.assertEqual(leafnode1.value, "textnode1") #"value" is not a known attribute of "None"
        self.assertIsNone(leafnode1.tag) #"tag" is not a known attribute of "None"
```

**Neovim, Lazy Package Manager**
I am running this project in Neovim with Lazy package manager with mason. All default configurations
```lua
pyright = {},
```
**Zip File**
[static_site_generator.zip](https://github.com/user-attachments/files/16200191/static_site_generator.zip)

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

The issue is that the type of `leafnode1` is `LeafNode | None`. As the error message indicates, you cannot access `value` on a `None` value.

I suspect that the reason PyCharm doesn't generate an error here is because they've probably hard-coded knowledge about `unittest.TestCase.assertIsInstance` in the type checker logic. Pyright is a standards-compliant type checker, and it bases its decisions only on type information provided to it in type stub files. It doesn't hard-code knowledge of specific libraries or modules.

If you want your code to type check without errors in pyright (or in mypy or in other standards-compliant type checkers), you'd need to add an assert statement before the point where you attempt to access the `value` attribute.

```python
assert leafnode1 is not None
```

For future bug reports, it would be helpful if you could provide a minimal, self-contained code snippet. In this case, it was relatively easy to diagnose the problem by downloading your full project, but often times this is very time consuming. And in the process of creating a minimal, self-contained repro, you can often diagnose the problem on your own.
